Adding RegNets to tf.keras.applications

AdityaKane2001 opened the pull request:
Please see #15240 and #15419. 

Progress so far:

X variant:

<table>
    <tr>
        <th>Model</th>
        <th>Paper (in %)</th>
        <th>Ours (in %)</th>
        <th>Diff (in %)</th>
        <th>Comments</th>
    </tr>
    <tr>
        <td>X002</td>
        <td>68.9</td>
        <td>67.15</td>
        <td>1.75</td>
        <td>adamw, area_factor=0.25</td>
    </tr>
    <tr>
        <td>X004</td>
        <td>72.6</td>
        <td>71.22</td>
        <td>1.38</td>
        <td>adamw, area_factor=0.08</td>
    </tr>
    <tr>
        <td>X006</td>
        <td>74.1</td>
        <td>72.37</td>
        <td>1.73</td>
        <td>adamw, area_factor=0.08</td>
    </tr>
    <tr>
        <td>X008</td>
        <td>75.2</td>
        <td>73.45</td>
        <td>1.75</td>
        <td>adamw, area_factor=0.08</td>
    </tr>
    <tr>
        <td>X016</td>
        <td>77</td>
        <td>75.55</td>
        <td>1.45</td>
        <td>adamw, area_factor=0.08, mixup=0.2</td>
    </tr>
    <tr>
        <td>X032</td>
        <td>78.3</td>
        <td>77.09</td>
        <td>1.21</td>
        <td>adamw, area_factor=0.08, mixup=0.2</td>
    </tr>
    <tr>
        <td>X040</td>
        <td>78.6</td>
        <td>77.87</td>
        <td>0.73</td>
        <td>adamw, area_factor=0.08, mixup=0.2</td>
    </tr>
    <tr>
        <td>X064</td>
        <td>79.2</td>
        <td>78.22</td>
        <td>0.98</td>
        <td>adamw, area_factor=0.08, mixup=0.3</td>
    </tr>
    <tr>
        <td>X080</td>
        <td>79.3</td>
        <td>78.41</td>
        <td>0.89</td>
        <td>adamw, area_factor=0.08, mixup=0.3</td>
    </tr>
    <tr>
        <td>X120</td>
        <td>79.7</td>
        <td>79.09</td>
        <td>0.61</td>
        <td>adamw, area_factor=0.08, mixup=0.4</td>
    </tr>
    <tr>
        <td>X160</td>
        <td>80</td>
        <td>79.53</td>
        <td>0.47</td>
        <td>adamw, area_factor=0.08, mixup=0.4</td>
    </tr>
    <tr>
        <td>X320</td>
        <td>80.5</td>
        <td>80.35</td>
        <td> 0.15</td>
        <td>adamw, area_factor=0.08, mixup=0.4</td>
    </tr>
</table>

Y variant:

<table>
<thead>
  <tr>
    <th></th>
    <th>Paper (in %)</th>
    <th>Ours (in %)</th>
    <th>Diff (in %)</th>
    <th>Comments</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>Y002</td>
    <td>70.3</td>
    <td>68.51</td>
    <td>1.79</td>
    <td>adamw, WD=1e-5, area_factor=0.16 mixup=0.2</td>
  </tr>
 <tr>
    <td>Y004</td>
    <td>74.1</td>
    <td>72.11</td>
    <td>1.99</td>
    <td>adamw, area_factor=0.16, mixup=0.2,WD=1e-5</td>
  </tr> 
 <tr>
    <td>Y006</td>
    <td>75.5</td>
    <td>73.52</td>
    <td>1.98</td>
    <td>adamw, area_factor=0.16, mixup=0.2</td>
  </tr> 
 <tr>
    <td>Y008</td>
    <td>76.3</td>
    <td>74.48</td>
    <td>1.82</td>
    <td>adamw, area_factor=0.16, mixup=0.2</td>
  </tr> 
 <tr>
    <td>Y016</td>
    <td>77.9</td>
    <td>76.95</td>
    <td>0.95</td>
    <td>adamw, area_factor=0.08, mixup=0.2</td>
  </tr> 
 <tr>
    <td>Y032</td>
    <td>78.9</td>
    <td>78.05</td>
    <td>0.85</td>
    <td>adamw, area_factor=0.08, mixup=0.2</td>
  </tr> 
  <tr>
    <td>Y040</td>
    <td>79.4</td>
    <td>78.2</td>
    <td>1.2</td>
    <td>adamw, area_factor=0.08, mixup=0.2</td>
  </tr>
  <tr>
    <td>Y064</td>
    <td>79.9</td>
    <td>78.95</td>
    <td>0.95</td>
    <td>adamw, area_factor=0.08, mixup=0.3</td>
  </tr> 
 <tr>
    <td>Y080</td>
    <td>79.9</td>
    <td>79.11</td>
    <td>0.69</td>
    <td>adamw, area_factor=0.08, mixup=0.3</td>
  </tr>
  <tr>
    <td>Y120</td>
    <td>80.3</td>
    <td>79.45</td>
    <td>0.85</td>
    <td>adamw, area_factor=0.08, mixup=0.4</td>
  </tr>
  <tr>
    <td>Y160</td>
    <td>80.4</td>
    <td>79.71</td>
    <td>0.69</td>
    <td>adamw, area_factor=0.08, mixup=0.4</td>
  </tr>
  <tr>
    <td>Y320</td>
    <td>80.9</td>
    <td>80.12</td>
    <td>0.78</td>
    <td>adamw, area_factor=0.08, mixup=0.4</td>
  </tr>
</tbody>
</table>

/cc @fchollet @sayakpaul @qlzh727 
/auto Closes #15240.

innat commented:
@AdityaKane2001 
Wondering, looks like it's the first time we're going to have lots of variants of the same model in `tf.keras.applications`. Great job anyway.

AdityaKane2001 commented:
@innat 

Yes, that is the case. Thank you :)

MrinalTyagi commented:
> @AdityaKane2001 Wondering, looks like it's the first time we're going to have lots of variants of the same model in `tf.keras.applications`. Great job anyway.

@innat Isn't addition of ResNet18 and 34 possible in similar fashion?

innat commented:
@MrinalTyagi 
I agree. But I don't know why these models aren't there. There're some requests [1](https://github.com/keras-team/keras/issues/15269), [2](https://github.com/keras-team/keras/issues/15270), [3](https://github.com/keras-team/keras/issues/15267), [4](https://github.com/keras-team/keras/issues/15268) are pending. Maybe there're some criteria that I'm unaware of.

lgeiger commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+                 group_width,
+                 stride=2,
+                 name=name + "_XBlock_0")(x)
+      for i in range(depth - 1):
+        x = XBlock(filters_out,
+                   filters_out,
+                   group_width,
+                   name=name + "_XBlock_" + str(i+1))(x)
```
```suggestion
      for i in range(1, depth):
        x = XBlock(filters_out,
                   filters_out,
                   group_width,
                   name=f"{name}_XBlock_{i}")(x)
```

lgeiger commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+                 group_width,
+                 stride=2,
+                 name=name + "_YBlock_0")(x)
+      for i in range(depth - 1):
+        x = YBlock(filters_out,
+                   filters_out,
+                   group_width,
+                   name=name + "_YBlock_" + str(i+1))(x)
```
```suggestion
      for i in range(1, depth):
        x = YBlock(filters_out,
                   filters_out,
                   group_width,
                   name=f"{name}_YBlock_{i}")(x)
```

lgeiger commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+                 filters_out,
+                 group_width,
+                 stride=2,
+                 name=name + "_XBlock_0")(x)
```
This code already uses f-strings in a few places, how do you feel about using them for things like this as well?
```suggestion
                 name=f"{name}_XBlock_0")(x)
```

lgeiger commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+                 filters_out,
+                 group_width,
+                 stride=2,
+                 name=name + "_ZBlock_0")(x)
```
```suggestion
                 name=f"{name}_ZBlock_0")(x)
```

lgeiger commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+                 group_width,
+                 stride=2,
+                 name=name + "_ZBlock_0")(x)
+      for i in range(depth - 1):
+        x = ZBlock(filters_out,
+                   filters_out,
+                   group_width,
+                   name=name + "_ZBlock_" + str(i+1))(x)
```
```suggestion
      for i in range(1, depth):
        x = ZBlock(filters_out,
                   filters_out,
                   group_width,
                   name=f"{name}_ZBlock_{i}")(x)
```

lgeiger reviewed (commented):
Thanks again for opening this PR! I have two tiny suggestions to slightly improve readability.

AdityaKane2001 commented:
@lgeiger  Always appreciated! Thanks for the changes. I'll merge them tomorrow.

mattdangerw commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+  def apply(inputs):
+    if filters_in != filters_out and stride == 1:
+      raise ValueError(
+          f"""Input filters({filters_in}) and output filters({filters_out}) are not equal for stride
```
Line too long. Conversion is...

```
raise ValueError(
          f"Input filters({filters_in}) and output filters({filters_out}) are not equal "
          f"for stride...
```

mattdangerw replied:
Fix other instances in this file as well.

mattdangerw commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+                   group_width,
+                   name=f"{name}_ZBlock_{i}")(x)
+    else:
+      raise NotImplementedError(f"""Block type `{block_type}` not recognized. 
```
Here as well, consecutive single quote strings.

mattdangerw commented on keras/applications/applications_load_weight_test.py:
```diff
@@ -69,7 +70,16 @@
         efficientnet_v2.EfficientNetV2L
-    ])
+    ]),
+    'regnet': (regnet,
```
I think you need a tf_py_test entry for this test in the BUILD file, you can see example for other load weights tests.

AdityaKane2001 replied:
I see.

mattdangerw commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+  if weights == "imagenet":
+    if include_top:
+      file_suffix = ".h5"
+      file_hash = WEIGHTS_HASHES[model_name[-2:]][0]
```
It looks like this part has not been tested? You are grabbing the last two characters, but your application keys above are all 4 characters long.

AdityaKane2001 replied:
Yes, actually. They are not tested yet. I'm still training the models.

mattdangerw commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+    }
+}
+
+BASE_DOCSTRING = """Instantiates Regnet architecture.
```
You need to actually use this. See https://github.com/keras-team/keras/blob/6c719b1a2229d554dd7714c22ff7a5a8a84e0ae7/keras/applications/efficientnet_v2.py#L1261 as an example.

mattdangerw commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+    name = "prestem" + str(backend.get_uid("prestem"))
+
+  def apply(x):
+    x = layers.Rescaling(scale=1. / 255., name=name + "_prestem_rescaling")(x)
```
Add an option for this `include_preprocessing=True` see efficientnet_v2 and mobilenet_v3

AdityaKane2001 replied:
I have added the option:
https://github.com/keras-team/keras/blob/6c719b1a2229d554dd7714c22ff7a5a8a84e0ae7/keras/applications/regnet.py#L803-L805

mattdangerw commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+      pooling=pooling,
+      classes=classes,
+      classifier_activation=classifier_activation
+  )
```
To follow out other applications, include `preprocess_input` and `decode_predictions` symbols. `preprocess_input` should just be a passthrough function as you are using rescaling in the model.

mattdangerw commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+      `classifier_activation=None` to return the logits of the "top" layer.        
+  
+  Returns:
+  A `keras.Model` instance.
```
indent, and add newline after returns block

mattdangerw commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+  Implementation of classification head of RegNet
+  
+  Args:
+  x: Input tensor
```
fix indent

mattdangerw commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+
+def SqueezeAndExciteBlock(filters_in, se_filters, name=None):
+  """
+  Implements the Squeeze and excite block(https://arxiv.org/abs/1709.01507)
```
space before paren

mattdangerw commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+                                  name=name + "_conv_3x3_bn")(x)
+    x = layers.ReLU(name=name + "_conv_3x3_relu")(x)
+
+    # Squeeze-Excitation block (https://arxiv.org/abs/1709.01507)
```
remove this comment, the link is in the function docstring already

mattdangerw commented on keras/applications/regnet.py:
```diff
@@ -0,0 +1,1549 @@
+                                  name=name + "_conv_3x3_bn")(x)
+    x = tf.nn.silu(x)
+
+    # Squeeze-Excitation block (https://arxiv.org/abs/1709.01507)
```
remove

mattdangerw reviewed (commented):
Thanks for the PR! Left a few comments. Do you have the weights for the applications available online somewhere?

AdityaKane2001 commented:
@mattdangerw 

Thanks for the review! 

> Do you have the weights for the applications available online somewhere?

I am still training these models, and I'm updating the [tables](https://github.com/keras-team/keras/pull/15702#issue-788479260) at the start of the thread accordingly. I'll test the loading code parallelly.

AdityaKane2001 commented:
@fchollet @mattdangerw 

I have completed training of all the models. I have updated the tables at the [start of the thread](https://github.com/keras-team/keras/pull/15702#issue-788479260) accordingly.  There are a couple of things I want to bring to your notice:

1. `model.predict(X_test)` does not work with grouped convolutions on CPU. For some reason, `model(X_test)` works flawlessly. Thus, I have updated the [`application_load_weight_test.py`](https://github.com/keras-team/keras/pull/15702/files#diff-89b6c8b391af2f95b405f44e3230796aaabf500cad8d49830d2a0fefba86c554R128) file accordingly.  If needed I'll open an issue for the same.
2.  All models are with 2% of the accuracies mentioned in the paper. Larger models are within 1%.

AdityaKane2001 commented:
@fchollet @mattdangerw 

Could you please take a look at this one?

mattdangerw commented:
@AdityaKane2001 thanks! Do you have the weights hosted online available to download somewhere? We will need that in order to test the model (and eventually upload the weights).

AdityaKane2001 commented:
@mattdangerw 

Yes, I have them in a GCS bucket: https://storage.googleapis.com/ak-regnet-savedmodels/

The `.h5` file paths are as expected, [`~/regnetx008.h5`](https://storage.googleapis.com/ak-regnet-savedmodels/regnetx008.h5) and [`~/regnetx008_notop.h5`](https://storage.googleapis.com/ak-regnet-savedmodels/regnetx008_notop.h5) for example.

mattdangerw reviewed (approved):
Looks great! Approving with two last comments.

Note that we will still need to go through a bit of process on our side to update weights, generate files for the new API and test everything out. If anything comes up there might need to message back here for changes.

Thank you!

AdityaKane2001 commented:
@mattdangerw 

I have made the requested changes. Please run the workflow again.

I wanted to provide performance metrics for all the models as on [keras.io/applications](https://keras.io/api/applications/). However, I am not able to spin up a VM instance of the given specs[^1] on Google Cloud. 

Could you please tell how to go about this?

/cc @sayakpaul 
[^1]: CPU: AMD EPYC Processor (with IBPB) (92 core) - Ram: 1.7T - GPU: Tesla A100

AdityaKane2001 commented:
@fchollet @mattdangerw

Could you please take a look at this one? TIA

AdityaKane2001 commented:
Thanks for the approval!

sayakpaul commented:
@mattdangerw could you also provide an update on what do we do about this? 

> I wanted to provide performance metrics for all the models as on keras.io/applications. However, I am not able to spin up a VM instance of the given specs1 on Google Cloud.

mattdangerw commented:
Yeah, re ideal machine for metrics, particularly the performance per step numbers, I am not sure. This is probably a question for @fchollet. We might take a bit to get back to you on this given it's the holidays and a lot of the team is out.

In the mean time, I think we can move ahead trying to land this PR, as the table update will be a separate change to keras.io anyway.

mattdangerw commented on keras/applications/applications_load_weight_test.py:
```diff
@@ -115,7 +125,7 @@ def test_application_pretrained_weights_loading(self):
       x = _get_elephant(model.input_shape[1:3])
       x = app_module.preprocess_input(x)
-      preds = model.predict(x)
+      preds = model(x).numpy()
```
Is there a reason this needs to be updated? Do things break otherwise? We still run these test cases in TF1 without eager mode enabled, and this line is breaking a number of our tests.

AdityaKane2001 replied:
@mattdangerw 

Yes, the change is necessary. Grouped convolutions are not yet fully supported on CPUs. We see that `model.predict(X_test)` breaks whereas `model(X_test)` works fine. 

There are a [number of issues](https://github.com/tensorflow/tensorflow/search?q=grouped+convolution&type=issues) discussing this in the TF repo.

mattdangerw replied:
I think you could trigger the failures by adding a call to `tf.compat.v1.disable_v2_behavior()` after before the call to `tf.test.main` in `applications_load_weight_test`. We can't submit this if we are breaking all these application tests in a TF1 context. We would need to find a change that does not rely on eager mode behavior (.numpy is eager only).

This might mean we need to dig into the difference between direct call vs predict here. It sound like this is an issue with grouped convolutions on CPU that will only appear when compiling a tf.function, is that right?

AdityaKane2001 replied:
@mattdangerw 

I have found a small solution to this. I have tested the change both with TF2 and using `tf.compat.v1.disable_v2_behavior()` and it works on my end. Could you please take a look and run the workflow again?

https://github.com/keras-team/keras/blob/851ca160f7c38e9effd9e60bcd44b6611073ae60/keras/applications/applications_load_weight_test.py#L127-L132

mattdangerw replied:
I will take a look next week! Sorry for the delayed reply, most of the team is out this week. The proposed change would still not submit, because that fallback (the numpy call) would still be run in a TF1 context for the regnet load weights test unless we disable it.

Overall, I think our options are...

 1) Disable the load weights test for regnet (without removing the predict call here), and follow up with a fix.
 2) Fix the underlying CPU/compiled function/grouped convolution issue, and then land this PR.
 3) Work around the bug for regnets somehow (the conversation [here](https://github.com/tensorflow/tensorflow/issues/51825) suggests that using `jit_compile=True` may allow CPU to work, which might give us a way forward).

I would say 3) would be the way to go if we can make it work. We really do want the load weights tests to test the compile predicted function (that's how these will often be used!), and shipping regnets such that predict will fail on CPU by default is not a great out of box experience.

Will follow up next week when people are back in office. Thanks!

AdityaKane2001 replied:
@mattdangerw 

Thanks for the detailed explanation. 

I guess (2) is not something which can be done in the Keras codebase, as error is thrown in [tensorflow/tensorflow/core/kernels/conv_ops_fused_impl.h](https://github.com/tensorflow/tensorflow/blob/1ef2bbae5f7521c9f1c3a831c162fe9a540df4b1/tensorflow/core/kernels/conv_ops_fused_impl.h#L216-L218).  I'll open an issue in the TF repo regarding this. So I agree that (3) might be the best option.

Lastly, wish you very happy new year!

AdityaKane2001 replied:
@mattdangerw 

Could you please take a look at this one? TIA

/cc @fchollet @qlzh727

mattdangerw replied:
Still here on this! We think we have found a good workaround (option 3), forcing XLA compilation grouped convolutions. https://github.com/keras-team/keras/pull/15868

Once that lands (assuming that doesn't run into road blocks), we can submit this without modifying the predict call in load weights tests.

AdityaKane2001 replied:
@mattdangerw 

Thanks a lot for this! Really appreciate it.

mattdangerw reviewed (commented):
I've uploaded weights and gen'd api files for this PR, things are looking overall good. However the change to applications_load_weight_tests is breaking right now. Commented on the line. Is that change necessary? Thanks!

AdityaKane2001 commented:
@mattdangerw 

Thanks a ton for #15868!

I have tested the code on my end and rolled back https://github.com/keras-team/keras/pull/15702/commits/851ca160f7c38e9effd9e60bcd44b6611073ae60 in https://github.com/keras-team/keras/pull/15702/commits/cf25748f546ef062d49e280cfe757c8d74f23f3e.

AdityaKane2001 commented:
Today these models were pushed to the [official docs](https://www.tensorflow.org/api_docs/python/tf/keras/applications/regnet). I sincerely thank the Keras team for allowing me to add these models. Huge thanks to the [TPU Research Group (TRC)](https://sites.research.google/trc/about/) for providing TPUs for the entire duration of this project, without which this would not have been possible. Thanks a lot to @fchollet for allowing this and guiding me throughout the process. Thanks to @qlzh727 for his guidance in building Keras from source on TPU VMs. Thanks to @mattdangerw for his support regarding grouped convolutions. Special thanks to @lgeiger for his contributions to the code. Last but not least, thanks a ton to @sayakpaul for his continuous guidance and encouragement.

mattdangerw commented:
Congrats on getting it landed and thanks for all the hard work on this! This is great to have!
